### ISSUE TITLE
[JENKINS-73] CVS changelog always shows edited files as added

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-06-a2#1549]
<p>All modified files (trunk or branch) in a CVS changes list are shown as added,<br/>
even when they really were modified. And as far as I can tell, files added on<br/>
branches are shown as edits. Files added in trunk appear correctly as adds.<br/>
Deletions in trunk also seem to be shown as adds.</p>

<p>The problem seems to be in the change log generator, not Hudson's parser, since<br/>
changelog.xml really shows things like:</p>

<p><entry><br/>
  <date>...</date><br/>
  <time>...</time><br/>
  <author><![CDATA<span class="error">&#91;...&#93;</span>]></author><br/>
  <file><br/>
    <name>.../X.java</name><br/>
    <revision>1.1.2.9</revision><br/>
  </file><br/>
  <msg><![CDATA<span class="error">&#91;...an edit; similar for a trunk commit...&#93;</span>]></msg><br/>
</entry></p>

<p>(note no <prevrevision>1.1.2.8</prevrevision>) or</p>

<p><entry><br/>
  <date>...</date><br/>
  <time>...</time><br/>
  <author><![CDATA<span class="error">&#91;...&#93;</span>]></author><br/>
  <file><br/>
    <name>.../X.java</name><br/>
    <revision>1.1.2.1</revision><br/>
  </file><br/>
  <msg><![CDATA<span class="error">&#91;...an add...&#93;</span>]></msg><br/>
</entry><br/>
<entry><br/>
  <date>...same...</date><br/>
  <time>...same...</time><br/>
  <author><![CDATA<span class="error">&#91;...same...&#93;</span>]></author><br/>
  <file><br/>
    <name>.../X.java</name><br/>
    <revision>1.1</revision><br/>
    <prevrevision>1.1.2.1</prevrevision><br/>
  </file><br/>
  <msg><![CDATA[branches:  1.1.2;<br/>
file X.java was initially added on branch xxx.]]></msg><br/>
</entry></p>

<p>(note synthetic add message, and <revision> vs. <prevrevision> backwards)</p>

<p><entry><br/>
  <date>...</date><br/>
  <time>...</time><br/>
  <author><![CDATA<span class="error">&#91;...&#93;</span>]></author><br/>
  <file><br/>
    <name>.../X.java</name><br/>
    <revision>1.2</revision><br/>
  </file><br/>
  <msg><![CDATA<span class="error">&#91;...useless now...&#93;</span>]></msg><br/>
</entry></p>

<p>(note no metadata indicating this was a deletion of file rev 1.1)</p>

---
<details><summary><i>Originally reported by <a title="Jesse Glick" href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-73" target="_blank">CVS changelog always shows edited files as added</a></i></summary>
<i><ul>
<li><b>reported by</b>: jglick
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>resolution</b>: Postponed
<li><b>resolved</b>: 2006-07-28T21:40:31+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 1
<li><b>imported</b>: 2025-11-06
</ul></i>
</details>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>Yeah, my colleages have noticied this but I haven't really paid serious<br/>
attention to this.</p>

<p>I believe CVS does offer enough information. Sounds like I need to hack the Ant<br/>
cvs changelog parser once again.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>CVS log output looks like the following:</p>

<p>----------------------------<br/>
revision 1.16<br/>
date: 2006/07/26 21:00:10;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:<br/>
kv;  commitid: 147c44c7d7da4eb2;  filename: foo.txt;<br/>
changed<br/>
----------------------------<br/>
revision 1.15<br/>
date: 2006/07/26 20:57:00;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:<br/>
kv;  commitid: 150044c7d71c4c46;  filename: foo.txt;<br/>
changed<br/>
----------------------------<br/>
revision 1.14<br/>
date: 2006/07/26 20:55:37;  author: kohsuke;  state: Exp;  lines: +1 -0;  kopt:<br/>
kv;  commitid: 166c44c7d6c84b34;  filename: foo.txt;<br/>
changed<br/>
=============================================================================</p>

<p>And '<prevrevision>' is obtained only when the "cvs log" command fetched more<br/>
than one entries from the server. The revision of the previous log record is<br/>
used as the 'previous revision' of the current log record.</p>

<p>Since Hudson runs "cvs log" with -d option, chances are, it only selects log<br/>
records that happened since the last build, and in that case, changelog.xml<br/>
doesn't record <prevrevision> because CVS didn't return any.</p>

<p>Therefore, it seems like after all it's not always possible to determine<br/>
add/edit/remove just from the CVS log command. We can distinguish remove from<br/>
other two by looking for "state: dead" line, but we can't distinguish add and<br/>
edit unless we check the previous revision.</p>

<p>I thought about using the output from "cvs update", but I still don't think we<br/>
can distinguish add and edit. I believe both may produce 'U' record.</p>

<p>Perhaps doing this correctly depends on issue #49. Then we can reliably find out<br/>
what data we are exchanging with the CVS server.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke">kohsuke</a>:</i>
<p>Meanwhile, I improved the parser to at least detect deletion.</p>

**Comment by jenkins-infra-bot** [comment]
<i><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick">jglick</a>:</i>
<p>Add vs. Edit is tricky because you can have a file 1.1 deleted in 1.2 and<br/>
readded in 1.3, so just checking rev.equals("1.1") will not always work. Also,<br/>
adding on a branch creates 1.1 as DEAD and a real rev 1.1.${2*k}.1 for some k>0,<br/>
and I presume readding a dead file on a branch produces some other form still.<br/>
So without a full file history it is hard to do more than guess. Since Hudson<br/>
now only asks the server for logs for files which really changed in the last<br/>
update, maybe it could leave off the -d arg to cvs log and try to extract the<br/>
important bits itself. Ugh.</p>

<p>Anyway I consider this a relatively low priority issue; it took me a long time<br/>
to even notice that document_add.gif was supposed to mean "added file". The file<br/>
name and log message are the important bits - if you really want to know what<br/>
changed you had better go find the diff, e.g. subscribe to the CVS change list<br/>
or have a ViewCVS server installed somewhere.</p>
